# Test Anything Protocol для Python

Данный модуль позволяет делать проверки и писать лог теста в TAP-формате


Пример теста:

```python

from easytap import Tap

t = Tap(3, 'Тест')
t.ok(foo(), 'foo вернул истину')
t.eq_ok(foo(), 123, 'foo вернул 123')

t.done_testing()

```

Особенности:

1. Планирует число проверок
2. Печатает лог в [TAP-протоколе](https://testanything.org/tap-specification.html)


## Чекеры/настройщики (API)

* `t.plan(cnt)` - Указывает системе план тестов в данном скрипте или сабтесте;
* `t.passed(msg)` - Указывает что тест с именем/описанием в `msg` пройден;
* `t.failed(msg)` - Указывает что тест с именем/описанием в `msg` не пройден;
* `t.ok(cond, msg)` - Проверяет условие на истину;
* `t.is_ok(v, expected, msg)` - Проверяет значение на соответствие (is) ожидаемому;
* `t.isnt_ok(v, unexpected, msg)` - Проверяет значение на не соответствие (is not);
* `t.in_ok(v, lst, msg)` - Проверяет что значение входит в список/кортеж/словарь/набор;
* `t.not_in_ok(v, lst, msg)` - Проверяет что значение НЕ входит в список/кортеж/словарь/набор;
* `t.eq_ok(v, expected, msg)` - Проверяет значение на соответствие (==) ожидаемому;
* `t.ne_ok(v, unexpected, msg)` - Проверяет значение на не соответствие (!=);
* `t.import_ok(module, msg)` - Проверяет может ли модуль быть импортирован;
* `t.isa_ok(value, class, msg)` - Проверяет является ли объект `value` членом класса class;
* `t.exception_ok(code, eclass, msg)` - проверяет на то что код выбросит исключение класса;
* `t.like(v, re, msg)` - проверяет на соответствие регекспу
* `t.unlike(v, re, msg)` - проверяет на НЕсоответствие регекспу

Для вывода диагностики можно использовать

* `t.note(msg)` - вывод в stdout (приводится к TAP)
* `t.diag(msg)` - вывод в stderr (приводится к TAP)

По завершению тестов можно вызвать `t.done_testing()` и в лог будет выведена
информация о статистике.

Этот же метод автовызывается из деструктора.

## Тестирование исключений

Если знаем что код должен выбросить исключение, то можем делать так:

```python

with tap.raises(RuntimeError, 'Выбрасывает исключение'):
    # тестируемый код
    raise RuntimeError('foobar')
```

## Сабтесты

Оформляются при помощи with:

```python

# тесты

with t.subtest(2, 'название блока сабтестов') as tap:
    tap.ok(foo(), 'foo выполнился')
    tap.eq_ok(bar(), 10, 'bar вернул 10')

```

**Внимание**: поскольку with не локализует переменную, может
возникать что-то мутное в случае если писать


```python
with t.subtest(2, 'название блока сабтестов') as t:
    ...
```

